core/templates/pages/exploration-player-page/services/hints-and-solution-manager.service.ts

```diff
@@ -16,190 +16,217 @@
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
 
-require(
-  'pages/exploration-player-page/exploration-player-page.constants.ajs.ts');
-
-angular.module('oppia').factory('HintsAndSolutionManagerService', [
-  '$timeout', 'PlayerPositionService',
-  'WAIT_FOR_FIRST_HINT_MSEC', 'WAIT_FOR_SUBSEQUENT_HINTS_MSEC',
-  function(
-      $timeout, PlayerPositionService,
-      WAIT_FOR_FIRST_HINT_MSEC, WAIT_FOR_SUBSEQUENT_HINTS_MSEC) {
-    var timeout = null;
-    var ACCELERATED_HINT_WAIT_TIME_MSEC = 10000;
-    var WAIT_FOR_TOOLTIP_TO_BE_SHOWN_MSEC = 60000;
-    var _solutionViewedEventEmitter = new EventEmitter();
-
-    var numHintsReleased = 0;
-    var numHintsConsumed = 0;
-    var solutionReleased = false;
-    var solutionConsumed = false;
-    var hintsForLatestCard = [];
-    var solutionForLatestCard = null;
-    var wrongAnswersSinceLastHintConsumed = 0;
-    var correctAnswerSubmitted = false;
-
-    var _hintConsumedEventEmitter = new EventEmitter();
-
-    // Variable tooltipIsOpen is a flag which says that the tooltip is currently
-    // visible to the learner.
-    var tooltipIsOpen = false;
-    // This is set to true as soon as a hint/solution is clicked or when the
-    // tooltip has been triggered.
-    var hintsDiscovered = false;
-    var tooltipTimeout = null;
-
-    PlayerPositionService.onNewCardAvailable.subscribe(
+import { EventEmitter } from '@angular/core';
+import { ExplorationPlayerConstants } from 'pages/exploration-player-page/exploration-player-page.constants.ts';
+import { Hint } from 'domain/exploration/HintObjectFactory';
+import { PlayerPositionService } from 'pages/exploration-player-page/services/player-position.service.ts';
+import { SubtitledHtml } from 'domain/exploration/SubtitledHtmlObjectFactory';
+import { Solution } from 'domain/exploration/SolutionObjectFactory';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class HintsAndSolutionManagerService {
+  timeout = null;
+  ACCELERATED_HINT_WAIT_TIME_MSEC: number = 10000;
+  WAIT_FOR_TOOLTIP_TO_BE_SHOWN_MSEC: number = 60000;
+  _solutionViewedEventEmitter = new EventEmitter();
+
+  numHintsReleased: number = 0;
+  numHintsConsumed: number = 0;
+  solutionReleased: boolean = false;
+  solutionConsumed: boolean = false;
+  hintsForLatestCard: Hint[] = [];
+  solutionForLatestCard: Solution = null;
+  wrongAnswersSinceLastHintConsumed: number = 0;
+  correctAnswerSubmitted: boolean = false;
+
+  _hintConsumedEventEmitter = new EventEmitter();
+
+  // Variable tooltipIsOpen is a flag which says that the tooltip is currently
+  // visible to the learner.
+  tooltipIsOpen: boolean = false;
+  // This is set to true as soon as a hint/solution is clicked or when the
+  // tooltip has been triggered.
+  hintsDiscovered: boolean = false;
+  tooltipTimeout = null;
+
+  constructor(private playerPositionService: PlayerPositionService) {
+    playerPositionService.onNewCardAvailable.subscribe(
```

Comment:
Can you create an issue called No subscriptions in services and put a TODO here?